### ISSUE TITLE
Describe headers using ABNF format

### ISSUE THREAD

**Issue by codefromthecrypt** [w3c/trace-context#81]
from @AloisReitbauer 

Honestly, I find it weird to describe the header format in plain English. We should use EBNF syntax to describe it. As an example look at the grammar for the HTTP Cookie header

...
cookie-date = *delimiter date-token-list delimiter
date-token-list = date-token ( 1delimiter date-token )
date-token = 1non-delimiter

delimiter = %x09 / %x20-2F / %x3B-40 / %x5B-60 / %x7B-7E
non-delimiter = %x00-08 / %x0A-1F / DIGIT / ":" / ALPHA / %x7F-FF
non-digit = %x00-2F / %x3A-FF

day-of-month = 1*2DIGIT ( non-digit *OCTET )
month = ( "jan" / "feb" / "mar" / "apr" /
"may" / "jun" / "jul" / "aug" /
"sep" / "oct" / "nov" / "dec" ) OCTET
year = 24DIGIT ( non-digit *OCTET )
time = hms-time ( non-digit OCTET )
hms-time = time-field ":" time-field ":" time-field
time-field = 12DIGIT

**Comment by AloisReitbauer** [comment]
Talked to Mark Nottingham today. He pointed me to http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html for header formats.

**Comment by AloisReitbauer** [comment]
Renaming issue to use "structured headers"

**Comment by AloisReitbauer** [comment]
I think the structured headers can be most easily used for Trace-State. We need to decide on which format we want to use:

* [Dictionary](http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#dictionary) - Which are basically unique name/value pairs. Example: _Example-DictHeader: foo=1.23, en="Applepie", da=*w4ZibGV0w6ZydGUK*_
* [Lists](http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#list) - Which is basciall an array of items _Example-StrListHeader: "foo", "bar", "It was the best of times."_
* [Parameterized Lists](http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#param). Which are a combination of both approaches. (Honestly, I find them confusing)

I think for our purpose dictionary is the best approach. 

I am, however, not sure how to represent Trace-Parent

**Comment by SergeyKanzhelev** [comment]
I added https://github.com/w3c/distributed-tracing/pull/96 for traceparent. 

I like how this document  http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html explains parsing algorithm. Would it be valuable to have the same for traceparent?

**Comment by SergeyKanzhelev** [comment]
Regarding `tracestate` I think Parameterized Lists is the best choice. It will allow `tc` without duplicating of entire `traceparent` **if** we will decide to go with this option.  Agree?

**Comment by SergeyKanzhelev** [comment]
Just to update this issue - we agreed on `Dictionary` during today's sync up. See [meeting notes](https://github.com/w3c/distributed-tracing/blob/master/meeting_notes/2018-04-10.md)

Need to copy boilerplate text from that spec to refer to `tracestate` as a dictionary. I promised to do it, but than @nikmd23 who sat in the same room said he wants to do it.

**Comment by yurishkuro** [comment]
> Just to update this issue - we agreed on Dictionary during today's sync up. See meeting notes

Sorry, I was not able to dial to the first half, and the notes do not provide a good explanation on that decision. I am not against it, but it creates two issues that the spec should address:
  1. how does the current node know that the previous node used the same or different tracer
  1. if the `tracestate` header grows too long and the current node must drop some of the other vendor entries, how does it choose which ones to drop?

Note that the positional format had very good answers to these questions.

**Comment by SergeyKanzhelev** [comment]
@yurishkuro the main concern with Parametrized List was items with the key without any value. The idea was to simplify the format. We haven't discussed ordering as mentioned in https://github.com/w3c/distributed-tracing/issues/80

You are correct. Dictionary implies unordered mapping:

> Let dictionary be an empty, unordered mapping.

So perhaps we can revisit that decision and agree on parametrized list

**Comment by codefromthecrypt** [comment]
>
> Just to update this issue - we agreed on Dictionary during today's sync
> up. See meeting notes
> <https://github.com/w3c/distributed-tracing/blob/master/meeting_notes/2018-04-10.md>
>
who attended this meeting? this is a standard. Considering the days of text
on the positional thing it is abysmal to obviate it with a one sentence
note.

@nicmunroe I should have made a bet that as soon as I stepped away
positional would return...though not sure anyone would have bet the other
side.

**Comment by SergeyKanzhelev** [comment]
> who attended this meeting? this is a standard. Considering the days of text on the positional thing it is abysmal to obviate it with a one sentence note.

Just to give a context - it wasn't long discussion as we mostly looked at format, not at parsing. So with parametrized list as was suggested earlier in this issue the confusion was around example `Example-ParamListHeader: abc_123;a=1;b=2; c, def_456, ghi;q="19";r=foo` with the mix of `;`-delimited and ','-delimited arguments (basically the format of `correlation-context` header).

We (at least I did) missed the keys positioning implication of the decision. As we follow the process with PRs approvals - it's great that @yurishkuro and you pointed that out. In any case we had an issue #80 opened for discussion and waiting for it's resolution.

If we want ParametrizedList, but without `;` - we will need to write our own spec for it. We can merge existing PR with the switch to Dictionary as it doesn't change the spec. And then by resolving issue #80 switch to ParametrizedList or custom definition of structured header.

Preferences?

**Comment by codefromthecrypt** [comment]
Preference is to stop changing things when no one has even implemented
current. Then your PRs to change things can come attached to N PRs of
impact.

It is well past due to have implementations and changing anything not
cemented by implementation is careless. While I might be missing something,
I have only seen one person ask to do an impl (a volunteer to zipkin4net).
Why are people holding meetings and still not cementing it in reality? Is
the draft really a draft or should we just consider it skippable as folks
holding meetings arent implementing it.

**Comment by SergeyKanzhelev** [comment]
@adriancole I do not quite understand what's the issue with switching to Dictionary. In the current shape of a spec there were no words about positioning. So it is basically no change. I understand it might be an issue if we would switch back to parametrized list and will have dictionaries of `;` delimited values with no order except for the first item. But even this I think reasonable change we can afford at this stage if it makes parsing easier (there are already well tested utilities for this in some platforms).

If you feel excluded due to meeting time - please comment/vote here https://github.com/w3c/distributed-tracing/issues/92

We failed to produce detailed notes from the last meeting. When I can I'll try to help making them more detailed.


> It is well past due to have implementations and changing anything not cemented by implementation is careless.

I can assure you I want the same. I can tell for myself while the delay. We are holding on AT LEAST final draft produced by Community Group. It doesn't mean we will not work on test cases as they'd help shape the requirements and find edge cases. Implementations requirement is in charter for working group and this is what we plan to satisfy.

This is what I put in "Status of This Document" of the report:
>There is a general agreement on names and format from the number of vendors. Community group is working in three directions:
> - Writing test cases to confirm viability of headers in practice.
> - Addressing edge cases scenarios.
> - Working on moving this effort forward with W3C by creating a working group.

**Comment by codefromthecrypt** [comment]
section 2.1 definitely conveys order. I think you know that

https://w3c.github.io/distributed-tracing/report-trace-context.html

anyway I don't have new information to offer based on your response.

**Comment by SergeyKanzhelev** [comment]
I want to resolve THIS issue. We have three options:
1. use Dictionary structured header. Implication - it implies no ordering of pairs. So it makes vague language into specific language. But it may need a change with the #80 
2. use Parameterized Lists structured header. It implies potentially parsing of extra "properties" after `;`. Also it makes vague language into specific language on ordering of key/value pairs. Again, it may need a change with #80.
3. Copy ABNF definition from structured header spec and remove ordering implication. Keep it vague until #80 will be resolved. 

@adriancole if you have strong preference to go with one option or another - please speak up.

Anybody else has opinion? I want to keep making progress and close issues one by one...

**Comment by yurishkuro** [comment]
I would prefer if our decisions are driven by semantics first, not by the format (or even worse, the _description of the format_) concerns. So here's a temporary solution for format definition to close this issue:
```
traceparent: TBD, pending resolution of https://github.com/w3c/distributed-tracing/issues/80
```

**Comment by codefromthecrypt** [comment]
The issue about how to markup in ABNF is separate from the order part.  I
don't have new information. I think the current spec, which is clearly
stack-based, is order based, and I don't want an accident of ABNF to remove
this.

If you all willingly choose to go with another route knowing the
implications I want you to do that loudly vs an accident of ABNF

**Comment by SergeyKanzhelev** [comment]
@adriancole it was an incident at first. Now we have a choice either to pick one with implication or to keep vague language for now. Currently published language of a format does not has any ordering specification. I'm surprised that structured headers spec does not define simple list. Maybe there is a hidden meaning in this I do not understand.

BTW, I wanted parametrized list in that discussion, not dictionary. So it was in-attentiveness to detail, not an attempt to hide something :-).

Anyway, let me refactor it to simple ABNF description without ordering. So we can close this issue and move on into #80.

**Comment by SergeyKanzhelev** [comment]
Another detail is that structured headers demand keys to be lowercase. I copied this requirement. Please comment if you think it's too much. I like this limitation though - makes think of short names.

**Comment by codefromthecrypt** [comment]
I like the lowercase feature. I think we've used it elsewhere for
consistency reasons.

**Comment by yurishkuro** [comment]
I'm good with lower case. The parameterized list sounds like the wrong notation for a number of reasons, especially the parameters after semicolons - we want the values to be opaque.

**Comment by SergeyKanzhelev** [comment]
Closed via #107

**Comment by nicmunroe** [comment]
I also like lowercase for keys, and I really don't like the idea of parameterized list with the combo of commas and semicolons (both for complexity and opaqueness reasons).

On the subject of which structured header format to use, why do we have to pick exactly one as it's officially defined, with no modifications? Why can't we say "tracestate is a [Dictionary](http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#dictionary) where order matters"? Or maybe "tracestate is a [List](http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#list) where each item in the list is a `key=value` pair with lowercase key"?

I'm sure I'm glossing over some subtleties that make it so you couldn't take either of my off-the-cuff examples above verbatim, but hopefully you get my point - could we do something like that, using an official structured header format as the "base" and then describing the small ways our header deviates?

**Comment by SergeyKanzhelev** [comment]
@nicmunroe I believe thinking here was that somebody already made heavy lifting of describing data structures and possibly have well tested implementations to parse those. So it might be easy to adopt one of them. So on surface it was easy sell.

Reading deep into spec of structured headers - it felt that `tracestate` is unique as it:
- requires ordering
- values are opaque (doesn't want to force putting quotes around opaque values treating them as strings)
- no desire to support numbers and booleans as values.
- no need to set additional attributes for values using `;` delimiter.

This https://github.com/w3c/distributed-tracing/pull/107 - has custom definition that worked around all those implications. I'd appreciate review as you did a great job finding issues in other PRs before.

Please note, that all these specifics lead to limitations with custom format:
- makes it harder to add extra property
- forcing opaqueness (not allowing numbers) makes it harder to compress values into binary format
- no properties associated with values makes impossible to mark up fields as PII (personal identifiable information) or implement other interesting features

So far I'm more under impression that custom format is better. It took some discussion to realize all implications of structured header format description.

**Comment by nicmunroe** [comment]
Gotcha on it being an easy win if we could pick something where there might already be tooling support. But if that means sacrificing important features of the spec then I'd rather abandon the existing-tooling desire in favor of a spec that is easy to understand and easy to write a parser for, even if you have to write it manually.

For example, an ordered dictionary seems like a pretty easy thing to understand and write a parser for with low likelihood of accidental bugs, even if you have to do it yourself.

I'm confused with some of the forced-opaqueness concerns though - defining the values as opaque just means the spec doesn't assign special meaning to anything in the value - i.e. the spec views each value as a black box, but the actual tracer/vendor generating a tracestate value can put anything in there. There's no restriction on numbers, or properties separated by semicolon, or any of that stuff. The only restriction is that the opaque value can't contain a comma - anything else is ok. Maybe I'm missing a potential benefit of making some of that stuff non-opaque?

**Comment by SergeyKanzhelev** [comment]
Structured headers requires quote around string. If no quotes it must be a number. With opaque values we do not want quotes at all

**Comment by nicmunroe** [comment]
Ah, that's what I get for skimming the structured headers doc. 

In that case I'm leaning towards custom format defined inline in the spec and keep it simple, even if that means a bit restrictive. If I have to go carefully read several sections of some other spec to not make a mistake (like I just did) then that's a warning sign IMO. At least for something that should be conceptually pretty simple.

**Comment by SergeyKanzhelev** [comment]
> Ah, that's what I get for skimming the structured headers doc.

exactly =) On a surface it was a very good idea to re-use.

**Comment by serialseb** [comment]
@SergeyKanzhelev Ditionaries, lists and param-lists in structured headers use header lists implicitly, so any usage of COMMA leads you to [RFC7230 Section 3.2.2](https://tools.ietf.org/html/rfc7230#section-3.2.2) which specifies ordering as being significant, without any need for further definition.

**Comment by AloisReitbauer** [comment]
The goal of the specification is to drive interoperability. This is not limited to tracer implementations but also middleware. The more we build on top of existing standards the easier it will be for other to adopt. The more we create a one of solution, the harder it will be.

IETF structured headers is still work in progress. I talked to @mnot and he is open to discuss. So let us put our requirements in place and have a joint discussion.

Regarding ordering: I am still not convinced that ordering is a good idea as it makes implementations less robust to faulty implementations. I also think we don't necessarily need it for anything. Last hop information can be stored in trace parent. My understand is, this is also what it is good for. As we then know what the last hop is, there is no more need for ordering. 

The goal of a specification is to build robust implementations on top of it. Ideally this includes robustness against implementation glitches of others.

**Comment by SergeyKanzhelev** [comment]
@AloisReitbauer yes, totally agree on reusing. One caveat here is that we need to make HTTP work for us, not make standard for for HTTP. We have other protocols to support. So less surprises we have - better standard it will become.

If the spec http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html can be adjusted for some restrictions we need to apply - it would be great. @mnot - as I mentioned above few differences we see are: ordering without the requirement to have secondary list elements (';'-delimited), ability to store opaque value without quotes and maximum length of entire set specification.

**Comment by SergeyKanzhelev** [comment]
@serialseb - yes, but structured headers RFC proposes to treat those as dictionaries.

**Comment by mnot** [comment]
@SergeyKanzhelev - I think you're looking for a list, correct? See: 
  http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#list

For opaque items, you can specify everything as strings, or specify that it can be any `item` type, or put things in binary blobs.

For maximum length, you can specify this yourself (note that the PR at https://github.com/httpwg/http-extensions/pull/578 is likely to land soon).

**Comment by SergeyKanzhelev** [comment]
@mnot we are looking for the list of `key`=`value` pairs. Implementing it as a list will make us express items of a list as strings - nothing else allows `=` sign in the middle. And will make us describe behavior of equal sign (`=`) and `OWS` surrounding it in our spec. So we will end up with 

```
tracestate: "key=value",  "key2 = value2"
```

Note - undesired quotes and we have a chunk of specification about the `key`-`value` expressed ourselves.

We do not want parametrized list due to unclear semantic of extra parameters (things delimited with `;`). Those might be fine with `Correlation-Context` header though. 

Also binary is not opaque - it is `base64`. And it requires surrounding `*`. Values like `1.5.3;29485735` or `app-id:{b6cee2e6-5b7b-476a-9cb0-b0224d1c9e41}` should be perfectly valid values without surrounding them with quotes or `*` sign.

Even examples of cookies given by [mozilla](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cookie) allows more flexibility in values: `PHPSESSID=298zf09hf012fh2; csrftoken=u32t4o3tb3gg43; _gat=1;`

Thank you for looking into this. If you believe our work can help you shape requirements for structured headers spec - I'll be happy to file issues or provide more details.

**Comment by mnot** [comment]
Is [this](https://w3c.github.io/distributed-tracing/report-trace-context.html#tracestate-field) the header field you're talking about?

If so, the closest thing is indeed a [Parameterised List](http://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#param). 

So, 

```
tracestate: rojo=00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01,
  congo=lZWRzIHRoNhcm5hbCBwbGVhc3VyZS4=
```
might become

```
tracestate: rojo;val="00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01",
  congo;val=*lZWRzIHRoNhcm5hbCBwbGVhc3VyZS4=*
```

... i.e., the tracing system name would be carried in the `identifier`, and the value in a parameter called `val` (for example). You could specify that other parameters are to be ignored (to allow extensions), or they could cause an error. Values could also be binary, etc.

Note that this does put some syntactic constraints on tracing system names.

> Values like 1.5.3;29485735 or app-id:{b6cee2e6-5b7b-476a-9cb0-b0224d1c9e41} should be perfectly valid values without surrounding them with quotes or * sign.

Why is this important? The value of using structured headers is that specification and parsing are taken care of for you, at least to a point where application-specific substructures come into play. If you want full control of the bytes on the wire, it's better not to use them, although you're much less likely to benefit from any more efficient encodings in the future.

**Comment by SergeyKanzhelev** [comment]
>Why is this important? 
I think the question is opposite - why would one want to surround opaque values with quotes. I understand the benefit may come for situations when parser need to quickly recognize whether the value is string or number. In case of this header - values are simply opaque blobs. So everything is a string. And consumer may have information on how to parse those blobs.

We will have live discussion today (tonight for me). I'll bring this for discussion.

**Comment by nicmunroe** [comment]
> Last hop information can be stored in trace parent. My understand is, this is also what it is good for. As we then know what the last hop is, there is no more need for ordering.

`traceparent` tells you what the trace state of the caller was (in `traceparent` format) - it does not tell you *who* the caller is. It's been mentioned before but I think bears repeating: knowing who the caller is can have significant benefit, even if it's "just" for logging/debugging purposes.

Ordering can also help us with logic for chopping `tracestate` down when it gets too large - i.e. we could recommend dropping the "oldest" values until the size is acceptable. Not that I'm suggesting that's what we *should* do for chopping logic, but with ordering we at least have the option.

**Comment by nicmunroe** [comment]
Reading the https://tools.ietf.org/html/rfc7230#section-3.2.2 section that @serialseb and others have brought up, in particular this portion:

> A sender MUST NOT generate multiple header fields with the same field
   name in a message unless either the entire field value for that
   header field is defined as a comma-separated list [i.e., #(values)]
   or the header field is a well-known exception (as noted below).

>   A recipient MAY combine multiple header fields with the same field
   name into one "field-name: field-value" pair, without changing the
   semantics of the message, by appending each subsequent field value to
   the combined field value in order, separated by a comma.  The order
   in which header fields with the same field name are received is
   therefore significant to the interpretation of the combined field
   value; a proxy MUST NOT change the order of these field values when
   forwarding a message.

Based on that, could we just define `tracestate` as a simple comma separated list, where each value is our expected `key=opaque` format? Something like this (syntax is probably wrong, but hopefully you get the idea):

```
tracestate = #(ts-value)
ts-value = key "=" opaque

[key and opaque defined appropriately to restrict characters, etc]
```

Then our definition for `tracestate` is succinct and easy to understand, *and* naturally follows the spec rules around multiple same-named headers, *and* gives significance to the ordering.

This is effectively what @SergeyKanzhelev put together for https://github.com/w3c/distributed-tracing/pull/107, just with a small tweak to more officially support the multiple-same-named-header stuff.

**Comment by mnot** [comment]
@nicmunroe that's an entirely reasonable approach. Much of the value of structured headers won't be realised/apparent until it's widely supported in tools (e.g., libraries, CDNs, Varnish, and so on). 

If you go this route, it would be good to define parsing as precisely as possible; e.g., is whitespace allowed around the `=`? what's the range of `opaque`? How do you encode a comma into `opaque`? What if `key` repeats? And so on.

**Comment by nicmunroe** [comment]
@mnot if you take a look at PR #107 I believe most (if not all) of those things you mention are specified in the changes, but I agree it should be defined as precisely as possible so we should double check all those points.

**Comment by SergeyKanzhelev** [comment]
Closed via #107
